lib/api/objectCopy.js

```diff
@@ -0,0 +1,306 @@
+                        dataStoreContext, log, (error, partRetrievalInfo) => {
+                            if (error) {
+                                return cb(error);
+                            }
+                            const partResult = {
+                                key: partRetrievalInfo.key,
+                                dataStoreName: partRetrievalInfo.dataStoreName,
+                                start: part.start,
+                                size: part.size,
+                            };
+                            return cb(null, partResult);
+                        });
+                    });
+                }, (err, results) => {
+                    if (err) {
+                        log.debug('error transferring data from source');
+                        return next(err);
+                    }
+                    return next(null, storeMetadataParams, results,
+                        destObjMD, serverSideEncryption);
+                });
+        },
+        function storeNewMetadata(storeMetadataParams, destDataGetInfoArr,
+            destObjMD, serverSideEncryption, next) {
+            return services.metadataStoreObject(destBucketName,
+                destDataGetInfoArr,
+                serverSideEncryption, storeMetadataParams, err => {
+                    if (err) {
+                        log.debug('error storing new metadata');
+                        return next(err);
+                    }
+                    // Clean up any potential orphans in data if object
+                    // put is an overwrite of already existing
+                    // object with same name
+                    // so long as the source is not the same as the destination
+                    let dataToDelete;
+                    if (destObjMD && destObjMD.location &&
+                        !sourceIsDestination) {
+                        dataToDelete = Array.isArray(destObjMD.location) ?
+                            destObjMD.location : [destObjMD.location];
+                        data.batchDelete(dataToDelete,
+                            logger.newRequestLoggerFromSerializedUids(
+                            log.getSerializedUids()));
+                    }
+                    return next(null, storeMetadataParams,
+                        serverSideEncryption);
+                });
+        },
+    ], (err, storeMetadataParams, serverSideEncryption) => {
+        if (err) {
+            return callback(err);
+        }
+        const xml = [
+            '<?xml version="1.0" encoding="UTF-8"?>',
+            '<CopyObjectResult>',
+            '<LastModified>', new Date(storeMetadataParams.lastModifiedDate)
+                .toUTCString(), '</LastModified>',
+            '<ETag>&quot;', storeMetadataParams.contentMD5, '&quot;</ETag>',
+            '</CopyObjectResult>',
+        ].join('');
+        const additionalHeaders = {};
+        if (serverSideEncryption) {
```

Comment:
may be we need to check here same as https://github.com/scality/S3/pull/146, to send back the key only for custom key id ?